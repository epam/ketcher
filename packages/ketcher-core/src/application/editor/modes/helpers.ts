import {
  BackBoneSequenceNode,
  EmptySequenceNode,
  HydrogenBond,
  LinkerSequenceNode,
  SubChainNode,
} from 'domain/entities';
import { ITwoStrandedChainItem } from 'domain/entities/monomer-chains/ChainsCollection';

export function isNodeRestrictedForHydrogenBondCreation(
  node: SubChainNode | BackBoneSequenceNode | undefined,
) {
  return (
    !node ||
    node instanceof LinkerSequenceNode ||
    node instanceof BackBoneSequenceNode ||
    node instanceof EmptySequenceNode
  );
}

export function isTwoStrandedNodeRestrictedForHydrogenBondCreation(
  twoStrandedNode?: ITwoStrandedChainItem,
) {
  const senseNodeHydrogenBonds =
    twoStrandedNode?.senseNode?.monomers.reduce((acc, monomer) => {
      return acc.concat(monomer.hydrogenBonds);
    }, [] as HydrogenBond[]) || [];

  return Boolean(
    isNodeRestrictedForHydrogenBondCreation(twoStrandedNode?.senseNode) ||
      isNodeRestrictedForHydrogenBondCreation(twoStrandedNode?.antisenseNode) ||
      twoStrandedNode?.antisenseNode?.monomers.some((monomer) =>
        monomer.hydrogenBonds.some((hydrogenBond) => {
          return senseNodeHydrogenBonds.includes(hydrogenBond);
        }),
      ),
  );
}
