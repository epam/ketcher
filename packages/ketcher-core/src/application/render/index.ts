export * from './raphaelRender'
export * from './restruct'
export * from './canvasExtension'
