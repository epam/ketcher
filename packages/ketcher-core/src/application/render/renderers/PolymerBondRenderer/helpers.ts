export const SMOOTH_CORNER_SIZE = 5;

export const generateBend = (
  dx1: number,
  dy1: number,
  dx: number,
  dy: number,
): string => {
  return `q ${SMOOTH_CORNER_SIZE * dx1},${SMOOTH_CORNER_SIZE * dy1} ${
    SMOOTH_CORNER_SIZE * dx
  },${SMOOTH_CORNER_SIZE * dy} `;
};

export const CORNER_LENGTH = 4;
export const DOUBLE_CORNER_LENGTH = CORNER_LENGTH * 2;

export const generateCornerFromTopToRight = () => {
  return `c 0,4.418 3.582,${CORNER_LENGTH} ${CORNER_LENGTH},${CORNER_LENGTH}`;
};

export const generateCornerFromLeftToTop = () => {
  return `c 4.418,0 ${CORNER_LENGTH},-3.582 ${CORNER_LENGTH},-${CORNER_LENGTH}`;
};

export const generateCornerFromBottomToRight = () => {
  return `c 0,-4.418 3.582,-${CORNER_LENGTH} ${CORNER_LENGTH},-${CORNER_LENGTH}`;
};

export const generateCornerFromBottomToLeft = () => {
  return `c 0,-4.418 -3.582,-${CORNER_LENGTH} -${CORNER_LENGTH},-${CORNER_LENGTH}`;
};

export const generateCornerFromLeftToBottom = () => {
  return `c 4.418,0 ${CORNER_LENGTH},3.582 ${CORNER_LENGTH},${CORNER_LENGTH}`;
};

export const generateCornerFromTopToLeft = () => {
  return `c 0,4.418 -3.582,${CORNER_LENGTH} -${CORNER_LENGTH},${CORNER_LENGTH}`;
};

export const generateCornerFromRightToTop = () => {
  return `c -4.418,0 -${CORNER_LENGTH},-3.582 -${CORNER_LENGTH},-${CORNER_LENGTH}`;
};

export const generateCornerFromRightToBottom = () => {
  return `c -4.418,0 -${CORNER_LENGTH},3.582 -${CORNER_LENGTH},${CORNER_LENGTH}`;
};
