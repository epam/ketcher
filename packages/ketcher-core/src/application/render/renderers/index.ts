export { ChemRenderer } from './ChemRenderer';
export { PeptideRenderer } from './PeptideRenderer';
export { BaseRenderer } from './BaseRenderer';
export { BaseMonomerRenderer } from './BaseMonomerRenderer';
export { PolymerBondRenderer } from './PolymerBondRenderer';
