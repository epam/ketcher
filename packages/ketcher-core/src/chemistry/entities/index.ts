export * from './Atom'
export * from './AtomList'
export * from './Bond'
export * from './Fragment'
export * from './HalfBond'
export * from './Loop'
export * from './RGroup'
export * from './RxnArrow'
export * from './RxnPlus'
export * from './SGroup'
export * from './SGroupForest'
export * from './SimpleObject'
export * from './Struct'
