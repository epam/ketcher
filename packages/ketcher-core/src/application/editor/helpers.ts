import { ketcherProvider } from 'application/utils';
import { BaseMode } from './modes/BaseMode';
import {
  DEFAULT_LAYOUT_MODE,
  HAS_CONTENT_LAYOUT_MODE,
  modesMap,
} from './modes';
import { KetSerializer } from 'domain/serializers';

export const initializeMode = (mode?: BaseMode): BaseMode => {
  if (mode) {
    return mode;
  }
  const ketcher = ketcherProvider.getKetcher();
  const isBlank = ketcher?.editor?.struct().isBlank();
  if (isBlank) {
    return new modesMap[DEFAULT_LAYOUT_MODE]();
  }
  return new modesMap[HAS_CONTENT_LAYOUT_MODE]();
};

export const parseMonomersLibrary = (monomersDataRaw: string | JSON) => {
  const monomersLibraryParsedJson =
    typeof monomersDataRaw === 'string'
      ? JSON.parse(monomersDataRaw)
      : monomersDataRaw;
  const serializer = new KetSerializer();
  const monomersLibrary = serializer.convertMonomersLibrary(
    monomersLibraryParsedJson,
  );

  return { monomersLibraryParsedJson, monomersLibrary };
};
