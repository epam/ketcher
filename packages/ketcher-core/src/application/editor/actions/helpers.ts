import { Atom, Bond, Neighbor, StereoLabel, Struct } from 'domain/entities';
import { StereoValidator } from 'domain/helpers';

export function getStereoAtomsMap(
  struct: Struct,
  bonds: Array<Bond>,
  bond?: Bond,
) {
  const stereoAtomsMap = new Map();
  const correctAtomIds: Array<number> = [];

  bonds.forEach((bond: Bond | undefined) => {
    if (bond) {
      const beginNeighs: Array<Neighbor> | undefined = struct.atomGetNeighbors(
        bond.begin,
      );
      const endNeighs: Array<Neighbor> | undefined = struct.atomGetNeighbors(
        bond.end,
      );

      if (
        StereoValidator.isCorrectStereoCenter(
          bond,
          beginNeighs,
          endNeighs,
          struct,
        )
      ) {
        const stereoLabel = struct.atoms.get(bond.begin)?.stereoLabel;
        if (
          stereoLabel == null ||
          stereoAtomsMap.get(bond.begin)?.stereoLabel == null
        ) {
          stereoAtomsMap.set(bond.begin, {
            stereoParity: getStereoParity(bond.stereo),
            stereoLabel: stereoLabel || `${StereoLabel.Abs}`,
          });
        }
        correctAtomIds.push(bond.begin);
      } else {
        if (!correctAtomIds.includes(bond.begin)) {
          stereoAtomsMap.set(bond.begin, {
            stereoParity: Atom.PATTERN.STEREO_PARITY.NONE,
            stereoLabel: null,
          });
        }
        if (!correctAtomIds.includes(bond.end)) {
          stereoAtomsMap.set(bond.end, {
            stereoParity: Atom.PATTERN.STEREO_PARITY.NONE,
            stereoLabel: null,
          });
        }
      }
    }
  });

  // in case the stereo band is flipped, changed or removed
  // TODO the duplication of the code below should be fixed, mayby by function
  if (bond) {
    if (!correctAtomIds.includes(bond.begin)) {
      stereoAtomsMap.set(bond.begin, {
        stereoParity: Atom.PATTERN.STEREO_PARITY.NONE,
        stereoLabel: null,
      });
    }
    if (!correctAtomIds.includes(bond.end)) {
      stereoAtomsMap.set(bond.end, {
        stereoParity: Atom.PATTERN.STEREO_PARITY.NONE,
        stereoLabel: null,
      });
    }
  }

  return stereoAtomsMap;
}

function getStereoParity(stereo: number): number | null {
  let newAtomParity: number | null = null;
  switch (stereo) {
    case Bond.PATTERN.STEREO.UP:
      newAtomParity = Atom.PATTERN.STEREO_PARITY.ODD;
      break;
    case Bond.PATTERN.STEREO.EITHER:
      newAtomParity = Atom.PATTERN.STEREO_PARITY.EITHER;
      break;
    case Bond.PATTERN.STEREO.DOWN:
      newAtomParity = Atom.PATTERN.STEREO_PARITY.EVEN;
      break;
  }
  return newAtomParity;
}
