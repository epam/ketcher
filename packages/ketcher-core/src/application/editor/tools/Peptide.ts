/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/
import { Tool } from 'application/editor/tools/Tool';
import { Peptide } from 'domain/entities/Peptide';
import { Vec2 } from 'domain/entities';
import { CoreEditor, fromPeptideAddition } from 'application/editor';
import { PeptideRenderer } from 'application/render/renderers/PeptideRenderer';
import { MonomerItemType } from 'domain/types';

class PeptideTool implements Tool {
  private peptidePreview: Peptide | undefined;
  private peptidePreviewRenderer: PeptideRenderer | undefined;
  readonly PEPTIDE_PREVIEW_SCALE_FACTOR = 0.4;
  readonly PEPTIDE_PREVIEW_OFFSET_X = 8;
  readonly PEPTIDE_PREVIEW_OFFSET_Y = 12;
  constructor(private editor: CoreEditor, private peptide: MonomerItemType) {
    this.editor = editor;
    this.peptide = peptide;
  }

  mousedown() {
    if (!this.peptidePreviewRenderer) {
      throw new Error('peptidePreviewRenderer is not initialized');
    }

    fromPeptideAddition(
      this.editor.renderersContainer,
      this.peptide,
      new Vec2(
        this.editor.lastCursorPosition.x -
          this.peptidePreviewRenderer.width / 2,
        this.editor.lastCursorPosition.y -
          this.peptidePreviewRenderer.height / 2,
      ),
    );
    this.editor.renderersContainer.update(false);
  }

  mousemove() {
    this.peptidePreview?.moveAbsolute(
      new Vec2(
        this.editor.lastCursorPosition.x + this.PEPTIDE_PREVIEW_OFFSET_X,
        this.editor.lastCursorPosition.y + this.PEPTIDE_PREVIEW_OFFSET_Y,
      ),
    );
    this.peptidePreviewRenderer?.move();
  }

  public mouseLeaveClientArea() {
    this.peptidePreviewRenderer?.remove();
    this.peptidePreviewRenderer = undefined;
    this.peptidePreview = undefined;
  }

  public mouseover() {
    if (!this.peptidePreview) {
      this.peptidePreview = new Peptide(this.peptide);
      this.peptidePreviewRenderer = new PeptideRenderer(
        this.peptidePreview,
        this.PEPTIDE_PREVIEW_SCALE_FACTOR,
      );
      this.peptidePreviewRenderer.show(this.editor.theme);
    }
  }
}

export { PeptideTool };
