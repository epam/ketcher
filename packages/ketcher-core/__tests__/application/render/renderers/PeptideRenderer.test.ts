import { createPolymerEditorCanvas } from '../../../helpers/dom';
import { PeptideRenderer } from 'application/render/renderers';
import { Peptide } from 'domain/entities/Peptide';
import { peptideMonomerItem, polymerEditorTheme } from '../../../mock-data';

describe('PeptideRenderer', () => {
  it('should render peptide', () => {
    const canvas = createPolymerEditorCanvas();
    const peptide = new Peptide(peptideMonomerItem);
    const peptideRenderer = new PeptideRenderer(peptide);
    global.SVGElement.prototype.getBBox = jest.fn();
    jest
      .spyOn(global.SVGElement.prototype, 'getBBox')
      .mockImplementation(() => ({ width: 30, height: 20 }));
    peptideRenderer.show(polymerEditorTheme);

    expect(canvas).toMatchSnapshot();
  });
});
