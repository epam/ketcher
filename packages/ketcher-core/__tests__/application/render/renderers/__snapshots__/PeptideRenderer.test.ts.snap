// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeptideRenderer should render peptide 1`] = `
<svg
  height="500"
  id="polymer-editor-canvas"
  width="500"
>
  <defs>
    <symbol
      id="peptide"
    >
      <path
        class="monomer-body"
      />
    </symbol>
  </defs>
  <g
    class="drawn-structures"
  >
    <g
      class="monomer"
      transform="translate(0, 0) scale(1)"
      transition="transform 0.2s"
    >
      <use
        fill="yellow"
        href="#peptide"
      />
      <text
        fill="black"
        font-size="6px"
        font-weight="700"
        line-height="6px"
        pointer-events="none"
        style="cursor: pointer; user-select: none;"
        x="0"
        y="10"
      >
        Abc
      </text>
      <text
        direction="rtl"
        fill="#7C7C7F"
        font-size="6px"
        font-weight="500"
        line-height="7px"
        text-align="right"
        x="10"
        y="-1"
      />
      <text
        direction="rtl"
        fill="#0097A8"
        font-size="6px"
        font-weight="700"
        line-height="7px"
        text-align="right"
        x="-6"
        y="10"
      />
    </g>
  </g>
</svg>
`;
